tests/block_sync_test.go

```diff
@@ -0,0 +1,179 @@
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package tests
+
+import (
+	"context"
+	"time"
+
+	"github.com/ethereum/go-ethereum/core/types"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+
+	"github.com/getamis/istanbul-tools/container"
+)
+
+var _ = Describe("Block synchronization testing", func() {
+	const (
+		numberOfValidators = 4
+	)
+	var (
+		blockchain container.Blockchain
+	)
+
+	BeforeEach(func() {
+		blockchain = container.NewBlockchain(
+			numberOfValidators,
+			container.ImageRepository("quay.io/amis/geth"),
+			container.ImageTag("istanbul_develop"),
+			container.DataDir("/data"),
+			container.WebSocket(),
+			container.WebSocketAddress("0.0.0.0"),
+			container.WebSocketAPI("admin,eth,net,web3,personal,miner"),
+			container.WebSocketOrigin("*"),
+			container.NAT("any"),
+			container.NoDiscover(),
+			container.Etherbase("1a9afb711302c5f83b5902843d1c007a1a137632"),
+			container.Mine(),
+		)
+
+		Expect(blockchain.Start()).To(BeNil())
+	})
+
+	AfterEach(func() {
+		Expect(blockchain.Stop(true)).To(BeNil())
+		blockchain.Finalize()
+	})
+
+	Describe("TFS-06 block synchronization testing", func() {
+		const numberOfNodes = 2
+		var nodes []container.Ethereum
+
+		BeforeEach(func() {
+			var err error
+			nodes, err = blockchain.CreateNodes(numberOfNodes,
+				container.ImageRepository("quay.io/amis/geth"),
+				container.ImageTag("istanbul_develop"),
+				container.DataDir("/data"),
+				container.WebSocket(),
+				container.WebSocketAddress("0.0.0.0"),
+				container.WebSocketAPI("admin,eth,net,web3,personal,miner"),
+				container.WebSocketOrigin("*"),
+				container.NAT("any"),
+			)
+
+			Expect(err).To(BeNil())
+
+			for _, n := range nodes {
+				err = n.Start()
+				Expect(err).To(BeNil())
+			}
+		})
+
+		AfterEach(func() {
+			for _, n := range nodes {
+				n.Stop()
+			}
+		})
+
+		It("TFS-06-01 node connection", func() {
+			By("Connect all nodes to the validators")
+			for i := 0; i < numberOfNodes; i++ {
+				nodeClient := nodes[i].NewIstanbulClient()
+				Expect(nodeClient).NotTo(BeNil())
+
+				for _, v := range blockchain.Validators() {
+					nodeClient.AddPeer(context.Background(), v.NodeAddress())
+				}
+
+				nodeClient.Close()
+			}
+
+			By("Wait for p2p connection")
+			<-time.After(10 * time.Second)
```

Comment:
A way to improve the speed of testing, we can poll blockchain.Validators() like every 100 millisecond and set a timeout for the whole `It`

Following will at most run for 10 seconds.
```
It("abc", func(done Done){
 close(done);
}, 10);
```